### ISSUE TITLE
HELP UBL !

### ISSUE THREAD

**Issue by ghost** [MarlinFirmware/Marlin#9596]
@thinkyhead @Roxy-3D 

Hi , UBL strange behavior outside of bedsize , up infinitly and more strange things, i have a purge bucket outside of the bed ' impossible to probe this void zone ' 

How to solve ?
1-) Enable/disable  UBL by gcode when go out of boundaries " when go on pause zone ' but i don't know if it can resume perfectly the height when back to print
2-) Move to the limit and go on pause zone , to avoid interpolation of bedzone and nobedzone
3-) Add manually the bucket zone in probing points
4-) Is it fixed now , Need support , or any options for out of probed zone  ?

Thanks

**Comment by Roxy-3D** [comment]
There have been a number of reports of the Z-Axis going to infinity on UBL.   And also a few reports of the Z-Axis diving into the bed.     I don't know when the regression got into the code.    I'm going to be making some additions to the UBL code so I'm going to be watching for this.  (AND FIX IT!!!!)

But the problem has to do with the nozzle going off the mesh.    You can do idea 1 and that will work.   UBL will 'resume perfectly the height' when it is re-activated.    You probably don't want to hear the reason this is...  But you can activate and de-activate as much as you want and you will not confuse the printer about the 'correct height'.

**Comment by ghost** [comment]
For information
- No bug when no moves in the bed and no big activity in the bed before go outside
- No bug when go on the limit , and after go outside
- The moves that go outside , stop a bed boundaries , up , and down
- No bug , when enable , disable

It's the interpolation calculation, when no probed points are available ' outside ' the maths are wrong for zheight
It's just here , the maths with no probed points
Add a rule ,  all over bedsize have the same value of the closest point

**Comment by Roxy-3D** [comment]
Agreed about the interpolation math.    The functions are coming up with a bad value if a NAN (invalid mesh point) is used in the calculations.   They should be coming up with 0.0 in that case.

**Comment by ghost** [comment]
I trust you

**Comment by ghost** [comment]
But , if you have a bed with a corner at 1mm
and you apply 0.0
The size of nozzle will shock the bed
It's why , you need to use the closest point to minimize the drama lol
Or create a define  like this  'UBL Outside z Height ' to have a fixed secure value for all outside movements  with disabled interpolation

**Comment by ghost** [comment]
UBL system force the user to activate an 'outside of bed ' secure height 
It's time to make it

**Comment by ghost** [comment]
Good job ! lol

**Comment by Roxy-3D** [comment]
Yes...  I understand why you are saying that.   But that is why it is very important that you setup and calibrate your printer.   If you tell the firmware the bed is a certain size....   Perhaps it would be good if the nozzle actually stayed in that area.    The problem with interpolating off of the nearest point (besides it taking a lot of calculations) is even that does not guarintee the nozzle won't be driven into the bed.

I'm more inclined to add an option something like:
```
#define UBL_Z_RAISE_WHEN_OFF MESH 5.00
```
If you really need to go off the mesh and UBL is active...   That number would be used as the correction instead of 0.000.      Would that be OK for your printer?

**Comment by studiodyne** [comment]
I want to give you my secret ....
I have created a mad purge bucket ... magic ... and cost nothing ...
This bucket will give the possibility to never use prime tower for dual extrusion.
But purge bucket must be outside the bed and it's why i need a known behavior when go to the bucket 
But if you ensure that the interpolation maths permit to disable/enable ubl when go to the bucket it's enough for me .
And you have all the time to fix this in the future.
I will try your new solution tomorow
Thks for help 
Promise the secret of my purge bucket xill be revealed soon .

**Comment by ghost** [comment]
Don't need new options 
Just add a comment on UBL description 

"To go outside of BEDSIZE enable/disable UBL" 

And ALL IS FIXED lol
And you have time to have reflexion about this and make some new maths

**Comment by ghost** [comment]
After reflexion and for the happyness of everybody ...

> #define UBL_Z_EXTERN_OFFSET 2    // mm

Is a very good idea
2mm is enough by default because feedrate of z never high and make time to raiseup

**Comment by Lutz1967** [comment]
Hi! First I would like to thank you for the great work. I learn to use Marlin now since some month and it`s great to apply and create. Recently I`ve tried to make UBL work on RAMPS. Firstly I had used MESH_INSET=15 to utilize the inductive probe correctly, but then I got the Z-to-the-max issue and also Z into the bed. Then I read that the root cause is going outside the mesh and I`ve redone everything with  MESH_INSET=0 and corrected the mesh by hand. Then I made a test print and at one point (clearly within the defined mesh) the extruder pushed out all the filament ... I have no idea how extruder and UBL interacts but that is what happened. 
Any idea?

**Comment by Roxy-3D** [comment]
I am currently working to resolve the Z-to-Infinity issue when the nozzle goes off the mesh.   

> Then I made a test print and at one point (clearly within the defined mesh) the extruder pushed out all the filament ... I have no idea how extruder and UBL interacts but that is what happened.
Any idea?

My guess is you have the filament sensor defined and there was some electrical noise on that signal.    If you are not really using the filament changer stuff, you might want to turn that off.

**Comment by ghost** [comment]
Me too , every day , the use of marlin is a pleasure , babystep + ubl + fw retract + advance is my happyness

**Comment by Lutz1967** [comment]
Thanks for the quick response. Look forward to use UBL when Z issue is fixed.

Filament runout stuff is not active. But I use the pin that is assigned to the sensor for something else (fan). I don’t think that should have any impact but I have corrected that.

**Comment by Athemis** [comment]
@Roxy-3D probably slightly off-topic, but perhaps it may help to pinpoint the Z-to-Infinity issue: I observed that issue on my delta printer not only if the nozzle goes off the mesh during UBL levelling, but also during regular prints. Those instances all had in common that the model was very close in size to the bed boundaries set in Marlin, though the model never really exceeded those (at least the slicer did not complain). But you can trigger that reproducably by trying to print gcode that has been sliced using bed boundaries slightly larger than those set in Marlin. My guess would be that some kind of rounding error is involved here?

**Comment by AnHardt** [comment]
Could xy-skew code be involved? (when leaving the probed area but being inside)

**Comment by Roxy-3D** [comment]
>  probably slightly off-topic, but perhaps it may help to pinpoint the Z-to-Infinity issue: I observed that issue on my delta printer not only if the nozzle goes off the mesh during UBL levelling, but also during regular prints. 

I was working on that all day today.   It is not a rounding issue.  It is a problem with invalid mesh points being used in the Z-Height correction calculations.    And the problem with going off the mesh is there is not another set of mesh points to do the bi-linear interpolation.

**Comment by ghost** [comment]
YEAHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHH

Roxy seems very close to push something soon
Don't forget to Force UBL to apply HEIGHT OF MESH OFF when  off mesh
Don't forget to add the ' definition of  ' HEIGHT OF MESH OFF ' ,  very very very very IMPORTANT
Don't forget to verify if babystep works  when disable
And the new release will be top
And tell me what units are impacted , the new bugfix have ' pause code ' totally destroyed ,and i don't want to use this new bugfix ,  so i will just update the units you have modified

_### Ps: For information , i have made a printing , dual color single nozzle , and each tool change need to go on purge bucket outside the bed , i have used G29 enable disable to go and back 
IT's perfect , there is just the beep to delete_

**Comment by Roxy-3D** [comment]
Right now I'm trying to fully understand the representation and usage of NAN in the Arduino library.   The reason is that is how the bad height is getting the stepper motors doing the wrong thing.   But something I don't understand is I have a configuration right now where doing a G26 causes the problem.  That is very strange (and others have reported this exact issue also) because G26 never sends the nozzle off the mesh.    

It will be a few days because I'm trying to fix this as cleanly as possible and not brute force the fix.  With that said...   I've got what I need now to get this resolved and I'm on it.

I'll probably do the Z_LIFT_WHEN_OFF_MESH as a separate Pull Request just because I have to change 37 Configuration.h files to do that one.

**Comment by Roxy-3D** [comment]
@studiodyne    I think the "Off the Mesh" issue is resolved so it doesn't make crazy up or down movements.    I need to check out the latest G26 changes for circle drawing...   And then I'll look at the other request to have a specified Z Height correction when off the mesh.

**Comment by Roxy-3D** [comment]
@studiodyne    The requested changes are merged into bugfix_2.0.0.    I'm making a bugfix_1.1.x version now....

https://github.com/MarlinFirmware/Marlin/pull/9661

Please let me know if works how you want it to work.   I'm wondering if the Z raise when the nozzle goes off the mesh should be 'faded' ?    At some height (maybe the fade height), it would seem the raise should not happen.   Right now, it raises unconditionally when the nozzle goes off the mesh.

**Comment by ghost** [comment]
YOU RULES !
I will test soon
Now , i'm working on the WANHAO LCD DLP , i just received it , at the software suit is so poor , two days and no print started
all is too young , and softwares are all prototypes , no smart function , for example , i want to print a tea cup , but to make economy of resin , i want to change the infill , with honey comb for example , and there is no software that transform a stl  with a hollowing/infill function ' and , the second question is ' how to get out the resin inside lolllllllllllll 
and more
Promise i test this soon

**Comment by filipgoc** [comment]
Just wanna say. Awesome troubleshooting @Roxy-3D!

Also happy to hear there's a regression or something in the mesh validating pattern G26. 
I presume now that's why babystepping does not work while G26ing. 

And I here I was thinking it was a _design choice_ (which I didn't understand). Haha.

**Comment by Roxy-3D** [comment]
My guess is if you enter Z-BabyStepping prior to doing the G26, it should work.   But really...   You want G26 to show you what the printer really sees.   The G26 pattern is not valuable if the user is moving the nozzle up and down.

**Comment by filipgoc** [comment]
@Roxy-3D 
I understand your reasoning, which I believe is based on starting with the printer being able to print in the center before even launching UBL. However, my use case is very different from yours.

Long story short, I have a predictable and consistent tilt to the build platform. Thus, it is way easier for me, if not necessary, to have some leveling in place and activated to be able print anything - and thus to be able to babystep the proper z probe offset.  Only after that I would move into validating the mesh.

(If you're wondering why I don't have the Z offset from the nozzle set in config as precisely as possible, it's because I have quick swap extruders with various hotends. It's way easier to lower or raise z probe into the right neighborhood when I swap out for a vastly different one, and then tweak the offset digitally in a few seconds.)

Does any of this make sense? 

I was poised to simply tweak the z offset as the validation pattern initiates (decidedly losing about a centimeter of validation) and then observe the rest of the pattern. Yet as I'm writing this, I suppose it would be indeed cleaner to add the z tweaking as an additional step prior to mesh validation...

I still wonder if it's necessary to _prevent_ user from using babystepping, if it actually is _by design._ If a user doesn't want to babystep, she or he doesn't have to. (User has to exhibit resolve, and act to babystep.)

**Comment by Roxy-3D** [comment]
> I still wonder if it's necessary to prevent user from using babystepping, if it actually is by design. If a user doesn't want to babystep, she or he doesn't have to. (User has to exhibit resolve, and act to babystep.)

Baby Stepping is done by constantly calling idle() and having the LCD code constantly checking the encoder wheel's position.   G26 doesn't do anything to make that happen.   But it could.   It probably is as easy as calling lcd_status_screen() when long waits are happening.   (And testing the system for correct operation.)     If you want to wake up G26 to baby stepping...   I'll help you get that to happen.

**Comment by filipgoc** [comment]
> If you want to wake up G26 to baby stepping...

Thanks. I want to keep you all focused on these other bugs for now :-) No distractions. 

I'll give it a bit more thought. Writing about this made me think that maybe what I want is adding a step to (my own) UBL flow.

**Comment by ghost** [comment]
How i use UBL 
110°c bed , i make a cold mesh :smile , i save , i set fade heigh to 1mm , and it's all
To verify if UBL works , i have a tape on the z axis ,if the tape move like a flag , all is good
To verify ubl by lcd , i have made a custom user script ' G1 xy move ' because , the LCD move does not act the ubl
I don't use the g26 pattern it's too long , time is precious
One click on the babystep ' option autosave ' , and ITS ALL
And every day , i think about roxy that have created this AWESOME MAGIC FEATURE !!!!!!!!!!!!

**Comment by ghost** [comment]
@Roxy-3D , please can you take one minute to read my post on your last pr conversation , i have made a little suggestion about ' Secure Z HEIGHT OFF MESh' , that works only when z axis is inferior to it , but when the z height have reached the secure value , you can disable this 
It's not good , not need , not usefull , increase the printing time ,and cost computation ticks ,to have a z raise during all the print 

Example
```
if (end[Z_AXIS]<=UBL_Z_RAISE_WHEN_OFF_MESH) z_raise = UBL_Z_RAISE_WHEN_OFF_MESH;
else  z_raise =0;

or

z_raise = (end[Z_AXIS]<=UBL_Z_RAISE_WHEN_OFF_MESH) ? UBL_Z_RAISE_WHEN_OFF_MESH : 0 ;
```
It's a feature to avoid scratching the bed , but this feature can be disabled when nozzle is far
it's a very light little fix 😸

**Comment by ghost** [comment]
UBL_Z_RAISE_WHEN_OFF_MESH 2.5
Not implemented in 2.0 just the definition exist in config.h

**Comment by Roxy-3D** [comment]
There was a problem with the use of #if ENABLED().    That had to be changed to #ifdef's because ENABLED() doesn't work with floating point numbers.     I'll try to bring up the bugfix_2.0.0 branch today and get it checked out (and working).

If you pulled the code before this commit:  https://github.com/MarlinFirmware/Marlin/commit/304e0f8945f8a25b9c08b5047dc8530a10f66975

Please grab a fresh copy and try it again...

**Comment by ghost** [comment]
Don't work anymore on 1.1.x now

**Comment by Roxy-3D** [comment]
Arrrrrgggghhhhhh!!!!

**Comment by ghost** [comment]
Not implemented too ?

**Comment by ghost** [comment]
I don't see the merged line ' calculations ' in the last bugfix 1.1.x

**Comment by ghost** [comment]
Travis not ok  loll

**Comment by Roxy-3D** [comment]
@Studiodyne    I like the idea that above the UBL_Z_RAISE_WHEN_OFF MESH nothing happens.   And below that height, it simply raises to that height.   I think we can accommodate that.   But that will take a little bit more work because right now a lot of the work is being done in the Z-Height correction routines and they don't know the current height.

**Comment by ghost** [comment]
Take your time
All the release versions are very bad now
I keep my old release
Today , i have made a new up , and  i have lost ' big edit font ' , ubl still in progress , pause system is shit
Marlin not on the highlights for the moments.
current z height is not like x and y , it's more calm and stable , and you can use the last known value in the routine with a secure margin , ultra precision not needed here , you just have to be up to the z value but never under

**Comment by ghost** [comment]
My wish for UBL
    off mesh Z secure height 'you know all '
    off mesh z secure active only under current z
    lcd babystepping active is UBL disabled ' seems to be solved '
    And a new wish ..... FADE HEIGHT in LCD , because fade height change when the object to print change , and there is no possiblity to adjust it by lcd , and if we print by sdcard , to set a fade height need usb connection and will reboot the board
    Fade height have been forgotten , but it's a VERY IMPORTANT DATA
    I use one millimeter fade height , because i d'ont understand why10mm
    if you print a plate of 3mm I understand that we want same thickness everywhere , but it's the only object that need a high fade height , all other object want to smooth the plan the more quick as possible

**Comment by Roxy-3D** [comment]
>  but it's the only object that need a high fade height , all other object want to smooth the plan the more quick as possible

No...  This isn't true.   The problem is the amount of extrusion isn't changed on a per layer basis.  The difference in the bed needs to be averaged into a large number of layers.

> And a new wish ..... FADE HEIGHT in LCD , because fade height change when the object to print change , and there is no possiblity to adjust it by lcd , and if we print by sdcard , to set a fade height need usb connection and will reboot the board

I believe it is there if you have Bed Leveling turned on in the LCD.
```
    #if ENABLED(ENABLE_LEVELING_FADE_HEIGHT)
      static float new_z_fade_height;
      void _lcd_set_z_fade_height() { set_z_fade_height(new_z_fade_height); }
    #endif
```

**Comment by AnHardt** [comment]
Even fade height could be dynamic.
If it needs 10 mm (50 layers) to even out a 1mm failure, that is 1/50 mm/layer.
If the largest failure (max difference of the probed points) on the bed is only 0.1mm it only needs 1mm (5 layers) fade height, to have the same (1/50 mm/layer) MAX_HEIGHT_CORRECTION.

**Comment by Roxy-3D** [comment]
> Even fade height could be dynamic.

Good point.   Fortunately...   This would be a relatively easy one to do within the SEGMENTED_MOVES logic because the firmware would know how much Z-Height correction was being applied in that area.    

> If it needs 10 mm (50 layers) to even out a 1mm failure, that is 1/50 mm/layer.

If we were to do this...  It might make sense to specify how much height is needed to fade out .1mm of error.    Some parts of the bed would not have any correction before other parts of the bed reached their fade height.

**Comment by Studiodyne** [comment]
The idea of dynamic is so perfect  , but , because of the extrusion not compensed all is impossible
Fuck !!! the idea is exactly what we want , no setting all automatic !
The good news is ' Lcd fade height is implemented ' , but i never found it ..
I go to find where it is implemented and in what menu
Thks !

**Comment by ghost** [comment]
@Roxy-3D 
The LCD FADE HEIGHT not implemented in the UBL menu
I've seen the ultralcd.cpp
No implementation , on the UBL zone
The menu is created , but never used
Make a fix ,  add it on the list lol ' easy ' 
I can make it , but i prefer when it's you , because MERGED immediatly , if it's me , i will be thinkyheadized before , and it will take a few days lolllllllllll

Updated wishes 
1-) Lcd Fade height for UBL 
2-) Nan/offmesh UBL bug
3-) Secure z height  ( + disable when current z raise to the same value ) 'Not important you have the time'

Thks in advance for the time you give here ! For us and for the firm

**Comment by ghost** [comment]
Updated wishes 
1-) Lcd Fade height for UBL 
2-) Secure z height  ( + disable when current z raise to the same value )
999/NAN -) Fix big edit font lollllllllllllllll PLEASE my favorite functionnality !!! 
Thks in advance for the time you give here ! For us and for the firm

**Comment by filipgoc** [comment]
For better or worse, this seems to be now temporarily a somewhat general UBL thread.

I am unclear about the use cases for **bed and hotend temperatures when probing the mesh.** 

- Is the ONLY reason to heat up the bed and the hotend before probing that most probes measure differently when heated up,  and the proximity to the hot bed (and extruder? really?)  is enough to mess them up? 
    - I suppose this is coupled with the bed actually varying in properties at different temperatures. 
- If that's the case, then what we really want to do is  _match_ is the temperature when generating the mesh and the temperature when homing as a part of our start gcode. 
- If I 
    - home all axes first
    - and then heat up the bed
    - and then print, ...
    - then I should actually be generating a _cold bed mesh_ even if I'm printing at 235/70 °C? 

Is this right?

**Comment by Roxy-3D** [comment]
Is the Bed Leveling main menu available when UBL is active?   If so, that is where it will be.
```
/**
 * Use the LCD controller for bed leveling
 * Requires MESH_BED_LEVELING or PROBE_MANUALLY
 */
//#define LCD_BED_LEVELING

#if ENABLED(LCD_BED_LEVELING)
```
I think the code is there.   I think we just need to get it to compile in when UBL is active.

**Comment by Roxy-3D** [comment]
> Is the ONLY reason to heat up the bed and the hotend before probing that most probes measure differently when heated up, and the proximity to the hot bed (and extruder? really?) is enough to mess them up?

Yes...   You will get different mesh values when probing cold or hot.   Personally...   I think one hot mesh is enough.   Because I use Z-BabyStepping when a print starts to manually adjust the first layer anyway.   But if you are not going to do that, you may want a mesh that is built at PLA temperatures and a mesh at ABS temperatures.   (And just save them to different mesh slots.)

I also don't really see a reason to generate a cold mesh.   But it is in the menu system...

**Comment by filipgoc** [comment]
> I also don't really see a reason to generate a cold mesh. But it is in the menu system...

Hm. What do you think of this line of reasoning? : 

> what we really want to do is match is the temperature when generating the mesh and the temperature when homing as a part of our start gcode.

If I home the z axis while the bed is cold and then heat up before the print, shouldn't I actually be generating a cold bed mesh even if I'm printing at 235/70 °C?

**Comment by ghost** [comment]
@filipgoc On a 400x400 buildplate or a 600x600 builplate , it's strongly requiered to probe with heated bed , because  the mesh isreally different , between 0.1/0.8mm , the hot bed mesh is not a joke , it's serious i have a 400mm bed , and if you have 0.4mm somewhere , your first layer is down and the nozzle scratch the bed , and i you have ' BUILDTAK' adhesive , you can put it in the trash !

@Roxy-3D  No available menu except the UBL big large strong infinite menu , so you can make a fix to have it

**Comment by filipgoc** [comment]
@Studiodyne

I think I got it. 

It's _not_ as much about the probe changing the probing (as that should mostly just shift the mesh up and down), but about the bed actually changing shape due to heating up? Is that what you're saying? 

(Otherwise, like roxy, just shifting the mesh up and down as needed...)

**Comment by ghost** [comment]
yes it changes shape , and the most common shape is because , the bed have fixation on the left and the right , the heat increase size of the bed , and force is applied on the fixation ,  the curve of the bed change , it's very light for eyes , but for a robotic nozzle absolutely perfect in an xy plan , you can be sure , in the center you have the nozzle that can not extrude , and in the corner  the filament not glue to the bed
It's why , you have to make the better probing .

**Comment by ghost** [comment]
it's why i just use 3 points matrix for ubl ,    one left , one center , one right , same for y axis , and it's enough , i don't have 32 bits computer so it's enough
and works REALLY PERFECT

**Comment by ghost** [comment]
The curse of the perfect first layer , is madness , it's why  i have my own limits !! lol

**Comment by filipgoc** [comment]
Right. I think for me it has not been as markant, as I don't have any machine with 400mm+ bed size and I use relatively thick aluminium (3+mm)... 

But of COURSE it makes sense. I was blindsided by thinking about the probe to the point that I forgot about the bed. 

Yet, this leaves the question of hotend.
- Why would we heat up the hotend? Because we presume proximity to the probe that would affect it (which is clearly very presumptuous)? Or because we are getting ready for G26? (In which case it'd be better to heat up later so that we are not baking an empty chamber.)

**Comment by Roxy-3D** [comment]
The hot ends expand when heated.   You can position your nozzle .05 mm above the unheated bed...   And when you turn on the nozzle heat, it is very possible you will see the nozzle touch the glass.

However you calibrate your machine, you should do the same thing (every time) when printing.    For me (and 80% or others) the preference is to get everything up to temperature, and then home and print.   But some people prefer a different sequence just because a print typically starts with a G28 (Home) and stuff can be heating while that is in operation.

**Comment by ghost** [comment]
@Filipgoc ,  in robotic 0.1mm is STRONG , so ,  all the closest conditions of printing must be reproduced for probing , and , we are all agree that it's really not difficult to heat nozzles and bed , for probing , it's why , we make it , and it's sure we will have a really better and more precise mesh
And no really a problem , because , i have made my mesh , a few month ago , and i still use it  today ,  the mesh is not deleted when you upload a new firmware

UBL is the autolevel that can have a medal of honor because EXTRAORDINARY AND EASY !
And this is  The @Roxy-3D  creation
Girlpower !

Updated wishes
1-) Lcd Fade height for UBL
2-) Nan/offmesh UBL bug
3-) Secure z height ( + disable when current z raise to the same value ) 'Not important you have the time'

**Comment by Roxy-3D** [comment]
> 2-) Nan/offmesh UBL bug

I believe this one is gone.

**Comment by ghost** [comment]
Coool 👍 

I test it soon , it's easy for me to test , i have a dual extrusion print , with purging in an offmesh bucket , every tool change ,  you imagine 250 tests in one print lol

**Comment by filipgoc** [comment]
@Studiodyne @Roxy-3D Thanks for helping clarify the variables and the context of the heating or no-heating for probing!

The sequencing is worth thinking about.
99% of time I'm printing PETG or PLA these days. 
I'l looking for a compromise then: 

I think what I'm gonna do in IN THE NAME OF UBL, all hail UBL, is this: 

Generate the mesh at 170/60. 
```
M104 S170   ; Hot enough so that leftover filament is soft and squished if necessary,
but no so much that it's oozing on my lovely PEI sheet. Mesh precision wise,
this doesn't matter because of babystepping. At this point we don't care. 
M190 S60    ; I don't normally print ABSes or polycarbonates, so this will
not be crazily off for by far most of my prints. 
```

Then, babystep at the very top of the range for PLA. 
```
M109 S225
; print something and adjust the z height to perfection
```

And combine with such start gcode
```
G28         ; home
(G0 X_ Y_    ; get the nozzle off-platform. in my case, safe homing does that already.)
M109 S170   ; heat up to the mesh temp, also nice to wait at lower temp while the bed
is heating so we're not baking the nozzle
M140 S60    ; heat up to the mesh temp (not sure if M190 is necessary here. thinking not)
G28 Z       ; Re-home Z for an updated value
M109 S{print temp} ; finish the heat up
M190 R{print temp} ; finish the heat up
; here, wipe nozzle if you have a nozzle wiper
; print
M117 Print starting
```

**Comment by Roxy-3D** [comment]
And I know people are tired of hearing it...    But Turning on the DOUBLECLICK_FOR_Z_BABYSTEPPING makes it much easier to get a perfect first layer independent of temperature.   As the printer is drawing the skirt,  You just adjust the nozzle so you are getting the right amount of 'squish' to line being put down.  

Probably it makes sense to have a perfectly calibrated machine for the filament you usually use.   But when you switch to another filament type for a print...   DOUBLECLICK_FOR_Z_BABYSTEPPING will make it so that print turns out OK.

**Comment by ghost** [comment]
My own idea of ubl , is to never be loud , so i use only one mesh , at a high temp , nozzle 250/bed 110 , and I never want to make 1200 meshes that will never change something
One mesh with heated machine and it's all and works perfect
Like you I use , pet and pla , and PCMAX , because these three filamentare odorless !!!!!!!!!!!!!!!!!!!!!!

M104 S250   
M190 S110
LCD 'make a cold mesh '  ' just a mesh to save , cold is not real '
LCD save on slot 0
LCD  SAVE CONFIG
And by usb  ' FADE HEIGHT =1 or any value you want ' **soon on LCD loll**
By usb too , i make long move g1 z0.5 ' because need to be under fade height '    g1 x1000y1000 to max the max long travel on the bed , WITH MY EYES , on the z axis to see correction

And it's all i do !
Do you think i'm enough mad to loose some hours just for making a pattern loll

For printing , i start the print , i double click i up to 3mm the baby stepping , ' GRAPHIC OVERLAY OPTION + AUTOSAVE OPTION ' 
When print begin , i set the height 
AND I  USE THE PRINT EVERY DAY and i never make another mesh since on month
 and i never touch babystepping , except when i change nozzle size

**Comment by ghost** [comment]
And every day i thank roxy !!!!!!!!!!!

**Comment by filipgoc** [comment]
Oh I love me babystepping and double click (I also have an alternative menu with Z babystepping prominent) and I am used to adjusting when skirt is along the way. 

But, I also like to have my printer quite far from me and on octoprint all the way until the print is done. It's a bit hard to see the first layer on a camera, so M290 wirelessly is of limited use. That's why I like to treat babystepping as a more of a setup procedure than a regular procedure. 

Idea!
I can make a mesh for every material (temp wise), save them to slots, and then add then load the meshes in the start gcode. 
I can make that part of the slicer to automatically switch gcode based on the material selected. Oh that's wicked. 

But it sounds a bit complicated for beginners (which has to be my first priority). Also, I have multiple beds - different surfaces and different properties. Once I save meshes for various temps and beds I am getting too many meshes to 'set and forget'. Then I'm getting back to generating the mesh on the fly and no worries (other than adding two minutes to my print start procedure.)

**Comment by ghost** [comment]
@filipgoc 
Don't make the curse of the perfection , this technology will never be perfect , and to loose time to have perfection is not serious ,  with this technology , when you make a print , you don't know if the same next print will be the same , it's why i never waste my time , i know this technology will never give high quality result , and i just require a succes first layer , and a good machining  and it's enough
I see tons of mad guys that give their life for a beautifull print , and when they see my printed object they are jalous et think i have a secret 
My secret is simple , the printed object must be cooled , to be hard and to keep dimensions , and all the closed machine are not good for me I ONLY USE OPEN MACHINES, in the past we loved ABS , and it's true we needed high temp  but now we know ABS IS SHIT and not made for printing , too much laminations and deformations
My second secret is the linear advance , you set it by make square , and you set to have the perfect corners
My third secret is FWRETRACT , to have the perfect recovery when the nozzle move and restart extrusion i set it in real time but , once you have the value , you never go back in settings
AND IT S ALL YOU CAN MAKE

**Comment by Roxy-3D** [comment]
> Also, I have multiple beds - different surfaces and different properties. Once I save meshes for various temps and beds I am getting too many meshes to 'set and forget'. 

So do it logically...  You know which materials you print with.   For me, if I was going to save multiple mesh's, I would do
- slot 0  mesh at room temperature
- slot 1  mesh at PLA temps (205)  bed at 60
- slot 2  mesh at ABS temps (225)  bed at 110
- slot 3  mesh at Nylon temp (240)  bed at 115

Also...  On my glass sheets I have markers (on the bottom side) so I can get the glass back very close to where the measurements were originally done.   When you have multiple pieces of glass for a printer, you can label those pieces of glass with the slot number for that particular piece of glass.

**Comment by ghost** [comment]
But if you love to make settings and love to test your machine , it's ok you can play lol

**Comment by ghost** [comment]
@Roxy-3D  can you tell me what units i have to add to have ubl nan/off mesh bug fixed , for my own release, because i don't want the new update , i will wait the new linear advance and other fixes

**Comment by filipgoc** [comment]
Once you have multiple quick swap beds, then the multiple meshes turn into

```
slot 0 mesh at room temperature,  bed 1
slot 1 mesh at PLA temps (205),   bed 1
slot 2 mesh at ABS temps (225),   bed 1
slot 3 mesh at room temperature,  bed 2
slot 4 mesh at PLA temps (205),   bed 2
slot 5 mesh at ABS temps (225),   bed 2
etc...
```

I'm not saying we should do something else. Just examining the limits of workflows. 

PS: I don't do this to find a perfection, but to make sure I'm setting up a properly thought out default workflow for the people that will learn how to do use UBL from me @Studiodyne. (Which is to say I agree with you in general. )

**Comment by ghost** [comment]
@filipgoc now , you have experience , and your job , is to transmit experience here , for the newbies
We have given time to you  , you have to promise , you will help newbies
👍

**Comment by Roxy-3D** [comment]
@filipgoc   If you run out of storage slots...   I'm not sure a mesh at room temp has much value.

@Studiodyne    I think it is likely that if you disable SEGMENT_LEVELED_MOVES that will be enough to make the problem go away.   But I added more code to make sure every path was covered.

**Comment by ghost** [comment]
What .h ?
What .cpp ?
I just want to update my own bugfix and i don't want to use a new updated release

**Comment by Roxy-3D** [comment]
it is in configuration.h

**Comment by ManuelMcLure** [comment]
I have a general question about UBL. I'm currently using bilinear leveling every time I start a print. I have a 220x220 bed (Anet A8) and an 18mm inductive sensor.

```cpp
#define FIX_MOUNTED_PROBE
#define X_BED_SIZE 220
#define Y_BED_SIZE 220
#define X_MIN_POS -36
#define Y_MIN_POS -7
#define Z_MIN_POS 0
#define X_MAX_POS X_BED_SIZE
#define Y_MAX_POS Y_BED_SIZE
#define Z_MAX_POS 240
#define AUTO_BED_LEVELING_BILINEAR
#define X_PROBE_OFFSET_FROM_EXTRUDER 71   // X offset: -left  +right  [of the nozzle]
#define Y_PROBE_OFFSET_FROM_EXTRUDER -35   // Y offset: -front +behind [the nozzle]
#define Z_PROBE_OFFSET_FROM_EXTRUDER -0.75 // Z offset: -below +above  [the nozzle]
#define GRID_MAX_POINTS_X 3
#define GRID_MAX_POINTS_Y GRID_MAX_POINTS_X
#define LEFT_PROBE_BED_POSITION 40
#define RIGHT_PROBE_BED_POSITION 200
#define FRONT_PROBE_BED_POSITION 20
#define BACK_PROBE_BED_POSITION 180
#define EXTRAPOLATE_BEYOND_GRID
```
Since I have an inductive probe, I make sure that the probe is at least 20mm in from the edge of the bed to ensure there's a nice chunky bit of aluminum under it, but the EXTRAPOLATE_BEYOND_GRID helps cover the cases outside the probable area giving me access to the whole bed.

Is there any way to get the same behavior using UBL? Essentially I want the mesh to cover the whole bed even though I'm only probing to within 20mm of the edge of the bed. If I set MESH_INSET to 20 I'll lose 20mm of printable area on each side of the bed, but if I set it to 0 it will want to probe the very edge of the bed which with an inductive probe is a problem. Should I just stay with Bilinear leveling for the moment? If this isn't possible, could there be an enhancement to add probe limits to G29 P1?

**Comment by Roxy-3D** [comment]
> Is there any way to get the same behavior using UBL? Essentially I want the mesh to cover the whole bed even though I'm only probing to within 20mm of the edge of the bed.

Yes.   This is no different than how a probe leg with a micro switch is handled.  If there is any X or Y_PROBE_OFFSET_FROM_EXTRUDER, you will end up with unprobed mesh points.   The simple answer is you do a 'Smart Fill' of them with a G29 P3.     You then do a G26 Mesh validation pattern.   and follow that with a G29 P4 to edit any spot that is not correct on the validation pattern.

**Comment by ManuelMcLure** [comment]
I understand that it takes into account the [XY]_PROBE_OFFSET_FROM_EXTRUDER and in my case that takes care of the probing on the left/back of the bed but I also want to make sure it doesn't probe the very edge of the bed on the front/right of the bed. The probe can physically reach those edges but is likely to be very inaccurate and trigger way too low (possibly even causing a nozzle crash) because there's not enough aluminum around it for proper inductive sensing. I want to tell UBL "don't probe all the way to the edge even if the probe can physically reach it" without using MESH_INSET and losing some of my build area, i.e. I want the completed mesh to go from [0,0] to [220, 220] (or maybe [5,5] to [215,215] for safety) but have G29 P1 only probe between (40, 20) and (200, 180).

However, a little digging into the code shows that you have already provided a way to do what I want, by just setting 
```cpp
#define MESH_MIN_X 40
#define MESH_MAX_X 200
#define MESH_MIN_Y 20
#define MESH_MAX_Y 180
```
in Configuration_adv.h. I somehow missed that in my earlier spelunking into the configuration of UBL. I think it might be time to give UBL another try (not that I have any problems with Bilinear).

**Comment by Lutz1967** [comment]
@Roxy-3D 

Is it OK to use MESH_INSET=20 and  disable SEGMENT_LEVELED_MOVES to get no Z issue outside the defined mesh ?

**Comment by Roxy-3D** [comment]
Yes.   Even though the nozzle can move across the entire bed...   Setting a MESH_INSET tells UBL not to allow the nozzle to go right up to the edge of the bed.   This is mostly for things like clamps to hold down a piece of glass to the bed.   By setting a MESH_INSET, you can make sure the nozzle won't hit any clamps.

SEGMENT_LEVELED_MOVES is currently ignored by UBL (because it was calling the wrong motion routine).   We will get that option back soon in UBL.  (It really is a powerful feature!!!)

**Comment by filipgoc** [comment]
As response to @ManuelMcLure ,

I would also love to know how to set it up like he wants it. My nozzle and probe can go anywhere on the bed, but that's just unrealistic to probe .5mm from the edge. I also believe that UBL does not extrapolate beyond the mesh boundaries like bilinear does, so...

_? Are we doomed _as of now_ to effectively make our corrected print area smaller by using UBL?_ 
(Compared to the bilinear ABL, as of now.)

I just did some testing: 

I flashed a firmware with insane INSETS **0mm**, but the MESH_MIN and MAX set to **70mm** from the edge. 

The probe does probe a small area in the middle, but then when observing the behavior with the mesh active, the areas off the middle were clearly not being compensated. 

(G29 A, G0 X0, G0 X180 F1500 and observe the compensation.)

So, it seems that MESH_MIN_X (and all the other overrides) are just way to adjust the INSET to be uneven, and resizing the  mesh smaller in the process.

**Comment by Roxy-3D** [comment]
> I would also love to know how to set it up like he wants it. My nozzle and probe can go anywhere on the bed, but that's just unrealistic to probe .5mm from the edge.

So...   go ahead and define the  full bed size.    If the points along the edge are not automatically probed accurately, don't sweat it.   You can edit them to be correct later when you do a G26 followed by a G29 P4 R

If the probe is going to do bad things when it tries to probe next to the edge, you can still do the same thing as up above.   But instead you start the process with a G29 P0 to clear the mesh.   And then manually probe or fill in the edge values with values that raise the nozzle enough you won't have a problem doing the G26.   Then auto probe the bed with G29 P1 C   and continue with the mesh refinement.

**Comment by ManuelMcLure** [comment]
That procedure seems a little unintuitive and prone to error compared to allowing the user to define a probable area smaller than the mesh. If I forget to probe the edges before running G29 P1 and I have a small Z offset I'm pretty likely to end up with my nozzle hitting the bed, since an inductive sensor at the edge of the bed will tend to trigger lower than it does when there's plenty of metal around it.

Would you be amenable to considering a feature request to allow UBL to honor [LEFT|RIGHT|FRONT|BACK]_PROBE_POSITION when doing G29 P1? If so I'll enter one.

**Comment by Roxy-3D** [comment]
I'm hesitant to add options for that.   If you set the bed size correctly, it will not probe off the bed.    If your probe doesn't work well near the edges of the bed, you can adjust the bed size to a size where the probe works and the nozzle can also reach all locations.

**Comment by ManuelMcLure** [comment]
> If your probe doesn't work well near the edges of the bed, you can adjust the bed size to a size where the probe works and the nozzle can also reach all locations.

But doesn't adjusting my bed size mean I'd lose printable area?

If so, it looks like my options are:
- Buy a different sensor that doesn't have problems near the edge of the bed
- Lose print area by saying that the bed is smaller
- Follow an error-prone procedure when scanning the bed to avoid nozzle crashes (and make sure I never forget to do it!)
- Use another leveling system

I think I'm going to take the last approach. Bilinear has been working well for me. I decided to test out UBL because I figured that it was where Marlin was moving towards in the future with respect with leveling and that it would get the most amount of development, but as it stands it's a big step backwards in ease-of-use for my specific circumstance (wanting to use the whole bed with an inductive probe).

**Comment by filipgoc** [comment]
As much as I LOVE UBL and think honestly it's the future of bed leveling, I am with Manuel here.

Without extrapolating the mesh beyond boundaries, it will be impossible to get a completely full access without  fussy procedures, at least with an inductive probe.*

And, since that is not an issue with a more 'primitive' leveling system, it can be seen as a step back - as much as UBL is overall a massive step forward. 

Let's focus on getting UBL stable first, then we can more it forward. 

\* (If we only get a full nozzle access by riding the probe to the very edge, then any slight misalignment or mis-sizing of beds, which in my case is quick swap for example, will mean a probing failure and driving the nozzle into the surface.)

**Comment by Roxy-3D** [comment]
> Without extrapolating the mesh beyond boundaries, it will be impossible to get a completely full access without fussy procedures, at least with an inductive probe

I've been hesitant to do that because I'm worried about Delta's cup and bowl problem.   At the Printable Radius, the curl of movement can be very extreme.    Perhaps we can make the extrapolation off the mesh only for Cartesian machines.  Or maybe make it a user selectable option where they can elect to turn it on.

**Comment by filipgoc** [comment]
@Roxy-3D Both are good points. As long as it is a config option, you should not worry too much about the cup and bowl. We can just add a warning for delta users to watch out. 

It doesn't have to propagate to LCD or anything... if there is a simple  `#define EXTRAPOLATE_BEYOND_GRID_UBL`, then everyone can make their choice according to their situation. 

I'd vote for extrapolation. 👍 This may be the one thing that still keeps some of us sticking with ABL bilinear.

**Comment by filipgoc** [comment]
As in: 

```c
#if ENABLED(AUTO_BED_LEVELING_UBL)

    // Extrapolate the mesh beyond the probed area? This is useful if you really
    // want to reach the ultimate extremities of your plate, but you are unable
    // or hesitant to probe that close to the edges. 
    // Default is to not apply any mesh compensation beyond the probed area. 
    // WARNING for Delta users. Use with extreme caution if you have any problems
    // of the cup & bowl variety. At the Printable Radius, the curl of movement can become extreme. 
    
    #define EXTRAPOLATE_BEYOND_GRID_UBL

**Comment by ghost** [comment]
@Roxy-3D  , yesterday , test of UBL , no offmesh bug , 200 moves out of mesh , zero problem ! But i have discover something , i have added the line to have zfadeheight with a MENUEDIT+float3+planner.zheight .Now i don't know if it's a little forgotten things or issue or need to be fixed , when someone make a LCD UBL , when mesh saved and UBL activated , fade height=0  no set to 10 'default' and we have to make it manually , but no LCD menu and forced to use ' USB '
The solution is to make a pr to have Z fade height menu that appears , because how many user will never see the z height compensation of their z axis ,  if fade height=0

Thks for this

**Comment by ghost** [comment]
@filipgoc 
It's true , the probing area is a real problem , in my situation i have made my machine with strong outside axis , and i can reach all the bed and more , but in the situation of already made printers , i suggest if you have bravery , to unmount your nozzle temporary , to place the sensor at the same place , and to make all your meshes  , and once you have your meshes , you have finish with this adventure

It's just once , and for life , you can give a little time to do it
:smile:

**Comment by AnHardt** [comment]
Wouldn't a additional ring of, not probable but editable, points around the current mesh avoid the extrapolation and replace it by meaningful data? These points could be preset by extrapolation.

**Comment by filipgoc** [comment]
> it's just once...

Well, it's once for you, but not for me and others like me who have quick-swap build plates. (Or many printers for that matter. The more printers, the more maddening this becomes.) Add into it that, ideally, as we had discussed, you'd re-generate mesh for vastly different print temperatures, and you are pretty far from 'once' ¯\_(ツ)_/¯
The discussion on adding some way to deal with this is underway in the meantime though, so I am happy.

**Comment by ghost** [comment]
quick swap okkkkkkkkkkk , it's every time lol
The quick swap bed is the magnetic buildtak thin plate  ? 
Give me a link i want to see what type of bed you have
i have forgotten the idea of a swapping bed , because of the ubl remeshing   
thks

**Comment by ghost** [comment]
@Roxy-3D  , The ' alien ' thinkyhead have found the bug of BIG EDIT FONT and have patched it ...

**Comment by Roxy-3D** [comment]
> Wouldn't a additional ring of, not probable but editable, points around the current mesh avoid the extrapolation and replace it by meaningful data? These points could be preset by extrapolation.

Yes.   And I've been trying to get this in place for Delta's.   In the Delta case, there naturally are extra mesh points outside of the Delta Printable Radius (except at the Max/Min X and Max/Min Y).   And it is my belief these need to be easily edited.

If we are willing to burn some extra RAM for these mesh points, that is a very viable solution.    BUT...   If we do that...   will people be coming back in two weeks saying "I can only go one mesh cell off of the bed...   That sucks and I don't want to use UBL any more."   ?????   Do we then go to surrounding the bed with 2 rows of mesh cells?

**Comment by filipgoc** [comment]
> "I can only go one mesh cell off of the bed... That sucks and I don't want to use UBL any more."

Hahaha. Unfortunately, maybe. Shoot for the 80/20 rule then? 

The main reason is to give easy way to provide compensation to the very edge of the plate. If one cell is enough to cover this depends on the density of probing and how far off edge we are probing.

Another reason could be to have compensation all the way to something like nozzle wiping gizmos or purge towers, but that should be fine without compensation I feel.

So... one cell should make it possible for many people to handle the edge state with automatic probes - given that they make their own choice on the density of the mesh and probing insets. There will be outliers. 2 cells would probably catch at least 80% of all cases, would be my guess. Beyond that... not sure.  

Maybe it still is a slippery slope of ram usage.

**Comment by Roxy-3D** [comment]
With 1 extra row (or column) of mesh cells...    The bed could be positioned such that the official area starts 1/2 into the first mesh cell.   And it would continue 1/2 into the last mesh cell of the row (or column).   That would be a minimal increase in RAM size.

But it maybe it would be more clean to have the extra mesh cells totally off the bed.   The big problem with doing that is those mesh points can not be auto probed.   They are off the bed.   They would all need to be manually edited to have meaningful data.   

It is possible we could have a user #define specify how many rows (or columns) of mesh cells off the bed are desired.   If they wanted 2, they could specify that.   But once again, that is a lot of manually entered data.

In the mean time...   Until this gets thought out better, it probably isn't that hard to get the code to extrapolate the nearest mesh value outwards off the mesh.   That would provide some relief while a better solution is figured out.   Probably it would make sense to make that user configurable.   
- no correction
- fixed height (which is now in place with #define UBL_Z_RAISE_WHEN_OFF_MESH 2.5)
- extrapolate nearest mesh point's value outward

**Comment by ManuelMcLure** [comment]
I think it still makes sense to have an option to limit the probable area so you can have actual mesh points near the edge of the bed that can be initialized with G29 P3 and fine tuned with G29 P4 without running the risk of probing too close to the edge of the bed.

In my ideal world, you'd specify X_BED_SIZE, Y_BED_SIZE, optionally MESH_INSET, [XY]_OFFSET_FROM_EXTRUDER, and the bed probe limits, and then:

- G29 P1 would probe only those mesh points that fall within the bed probe limits
- G29 P3 would fill in the mesh points outside the probe limits with reasonable values
- G26 + G29 P4 would allow fine tuning of the points.
